File: lingting-core/src/main/java/live/lingting/framework/s3/S3Properties.java
package live.lingting.framework.s3;

import lombok.Getter;
import lombok.Setter;

/**
 * @author lingting 2024-09-12 21:20
 */
@Getter
@Setter
public class S3Properties {

	protected String scheme = "https";

	protected String region;

	protected String endpoint;

	protected String bucket;

	protected Acl acl = Acl.PRIVATE;

	protected String ak;

	protected String sk;

	protected String token;

	public <T extends S3Properties> T fill(T properties) {
		properties.setScheme(getScheme());
		properties.setRegion(getRegion());
		properties.setEndpoint(getEndpoint());
		properties.setBucket(getBucket());
		properties.setAcl(getAcl());
		properties.setAk(getAk());
		properties.setSk(getSk());
		properties.setToken(getToken());
		return properties;
	}

	public S3Properties copy() {
		S3Properties properties = new S3Properties();
		fill(properties);
		return properties;
	}

	public void useCredential(Credential credential) {
		setAk(credential.getAk());
		setSk(credential.getSk());
		setToken(credential.getToken());
	}

}

File: lingting-aws/src/main/kotlin/live/lingting/framework/aws/s3/properties/S3Properties.kt
package live.lingting.framework.aws.s3.properties

import live.lingting.framework.aws.policy.Acl
import live.lingting.framework.aws.policy.Credential
import live.lingting.framework.aws.s3.AwsS3Properties
import live.lingting.framework.aws.s3.enums.HostStyle
import live.lingting.framework.http.HttpUrlBuilder

/**
 * @author lingting 2025/1/14 17:14
 */
abstract class S3Properties {

    open var ssl: Boolean = true

    open var hostStyle: HostStyle = HostStyle.VIRTUAL

    open var region: String = "us-east-1"

    open var endpoint: String = "amazonaws.com"

    open var bucket: String = ""

    open var acl: Acl = Acl.PRIVATE

    open var ak: String = ""

    open var sk: String = ""

    open var token: String? = ""

    open fun useCredential(credential: Credential) {
        ak = credential.ak
        sk = credential.sk
        token = credential.token
    }

    open fun from(properties: S3Properties) {
        ssl = properties.ssl
        region = properties.region
        endpoint = properties.endpoint
        bucket = properties.bucket
        acl = properties.acl
        ak = properties.ak
        sk = properties.sk
        token = properties.token
    }

    open fun copy(): S3Properties {
        return AwsS3Properties().also { it.from(this) }
    }

    open fun host(): String {
        if (hostStyle == HostStyle.VIRTUAL) {
            return virtualHost()
        }
        return secondHost()
    }

    open fun urlBuilder(): HttpUrlBuilder {
        val builder = HttpUrlBuilder.builder()
            .host(host())

        if (ssl) {
            builder.https()
        } else {
            builder.http()
        }

        if (hostStyle == HostStyle.SECOND) {
            builder.path(bucket)
        }

        return builder
    }

    open fun virtualHost(): String {
        val host = secondHost()
        if (bucket.isNotBlank()) {
            return "$bucket.$host"
        }
        return host
    }

    abstract fun secondHost(): String

}
